### PR TITLE
[Wholegraph] Add wholgraph format conversion script

### PR DESCRIPTION
**Pull request by isratnisa**
This PR adds the script to convert the node features (`'feat'`) and labels from the `node_feat.dgl` file into a format compatible with [WholeGraph](https://github.com/rapidsai/wholegraph). These converted features are then saved as `.../wholegraph/<ntype>~<feat>`. The `node_feat.dgl` file is subsequently trimmed to remove the feature and label attributes, leaving only other attributes such as `train_mask`, `test_mask`, and `val_mask` intact. 

Users must convert the features to leverage the WholeGraph framework to accelerate node feature transfer across multiple machines ([PR](https://github.com/awslabs/graphstorm/pull/405)).

The script and README are developed by Nicolas Castet  (@nvcastet) from NVIDIA.

### CODE REVIEW

File: tools/convert_feat_to_wholegraph.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        print(f"Processing '{f}' features...")
+        if f not in node_feats_data[0]:
+            print(f"Error: Unknown feature '{f}'. Files contain the following features: {node_feats_data[0].keys()}.")
+            sys.exit(1)
+        whole_feat_tensor = torch.concat(tuple(t[f] for t in node_feats_data), dim=0)
+        metadata[f] = {'shape': list(whole_feat_tensor.shape), 'dtype': str(whole_feat_tensor.dtype)}
+        # Round up the integer division to match WholeGraph partitioning scheme
+        subpart_size = -(whole_feat_tensor.shape[0] // -num_parts)
+        for part in range(num_parts):
+            l = part*subpart_size
```
Comment (classicsong) [inline tools/convert_feat_to_wholegraph.py:40]: Please use more meaningful variable name.

File: tools/convert_feat_to_wholegraph.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+            u = (part+1)*subpart_size if part != (num_parts - 1) else whole_feat_tensor.shape[0]
+            wg_tensor = wgth.create_wholememory_tensor(local_comm, 'continuous', 'cpu', (u-l, *whole_feat_tensor.shape[1:]), whole_feat_tensor.dtype, None)
```
Comment (classicsong) [inline tools/convert_feat_to_wholegraph.py:42]: Line exceeds 100.

File: tools/convert_feat_to_wholegraph.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+    # Save new truncated distDGL tensors
+    for part in range(num_parts):
+        dgl.data.utils.save_tensors(os.path.join(folder, f'part{part}', 'new_node_feat.dgl'), node_feats_data[part])
+
+    # swap 'node_feat.dgl' files
+    for part in range(num_parts):
+        os.rename(os.path.join(folder, f'part{part}', 'node_feat.dgl'), os.path.join(folder, f'part{part}', 'node_feat.dgl.bak'))
+        os.rename(os.path.join(folder, f'part{part}', 'new_node_feat.dgl'), os.path.join(folder, f'part{part}', 'node_feat.dgl'))
+
```
Comment (classicsong) [inline tools/convert_feat_to_wholegraph.py:65]: Please add a TODO(edge feature)

File: tools/README.md [hunk 4]
```diff
@@ -83,3 +83,51 @@ python3 /graphstorm/tools/partition_graph.py --dataset ogbn-arxiv \
...
+                    global_comm,
+                    'distributed',
+                    'cpu',
+                    getattr(torch, metadata['paper/feat']['dtype'].split('.')[1]),
+                    metadata['paper/feat']['shape'],
+                    optimizer=None,
+                    cache_policy=cache_policy,
+                )
+# 'part_count' is the number of partition files. For distDGL it will always be the number of machines.
+paper_feat_wg.get_embedding_tensor().from_file_prefix('ogbn-mag240m/wholegraph/paper~feat', part_count=4)
```
Comment (chang-l) [inline tools/README.md:131]: TBH, I don't know why we need to split wholegraph features into 4 different files, and save all under one folder `ogbn-mag240m/wholegraph`... It would not save any space, right?

File: tools/README.md [hunk 5]
```diff
@@ -83,3 +83,51 @@ python3 /graphstorm/tools/partition_graph.py --dataset ogbn-arxiv \
                                              --output /tmp/ogbn_arxiv_nc_train_val_1p_4t  \
                                              --is-homo     
 ```
+
+## Convert node features from distDGL format WholeGraph format
+
+Use the `convert_feat_to_wholegraph.py` script with `--dataset-path` pointing to the distDGL folder of partitions and `--feat-names` to the features you want to convert (e.g. '`paper/feat,author/feat,institution/feat`')
+```
+python3 convert_feat_to_wholegraph.py --dataset-path ogbn-mag240m-2p --feat-names paper/feat
```
Comment (zheng-da) [inline tools/README.md:91]: let's not expose internal node data names. you should use the same convention as training script. I believe the conversion is something like "paper:feat"

File: tools/README.md [hunk 6]
```diff
@@ -83,3 +83,51 @@ python3 /graphstorm/tools/partition_graph.py --dataset ogbn-arxiv \
                                              --output /tmp/ogbn_arxiv_nc_train_val_1p_4t  \
                                              --is-homo     
 ```
+
+## Convert node features from distDGL format WholeGraph format
```
[the remark this answers is not in the document]
Comment (isratnisa) [reply tools/README.md:87]: Fixed

File: tools/README.md (hunk 4 above)
Comment (isratnisa) [reply tools/README.md:131]: No it does not save space. The dataset is partitioned in 4 parts. Are you suggesting to put them in separate folder? Originally, the feats will be in 
```
ogbn-mag240m/part0/node_feat.dgl 
ogbn-mag240m/part1/node_feat.dgl
...
```
now the the features are in 
```
ogbn-mag240m/wholegraph/paper~feat_part_0_of_4
ogbn-mag240m/wholegraph/paper~feat_part_1_of_4
...
```

File: tools/convert_feat_to_wholegraph.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        # Round up the integer division to match WholeGraph partitioning scheme
+        subpart_size = -(whole_feat_tensor.shape[0] // -num_parts)
+        for part in range(num_parts):
+            l = part*subpart_size
+            u = (part+1)*subpart_size if part != (num_parts - 1) else whole_feat_tensor.shape[0]
+            wg_tensor = wgth.create_wholememory_tensor(local_comm, 'continuous', 'cpu', (u-l, *whole_feat_tensor.shape[1:]), whole_feat_tensor.dtype, None)
+            local_tensor, _ = wg_tensor.get_local_tensor(host_view=True)
+            local_tensor.copy_(whole_feat_tensor[l:u])
+            filename = wgth.utils.get_part_file_name(f.replace('/', '~'), part, num_parts)
+            wg_tensor.local_to_file(os.path.join(wg_folder, filename))
```
[the remark this answers is not in the document]
Comment (isratnisa) [reply tools/convert_feat_to_wholegraph.py:46]: We decided to support multiple features. We will count anything other than `*mask` and `labels` as features. And yeah it will be messy...

File: tools/README.md (hunk 4 above)
Comment (chang-l) [reply tools/README.md:131]: Yes, saving them in separate folders is better as it allows parallel data distribution in clusters.

File: tools/convert_feat_to_wholegraph.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    for path in (os.path.join(folder, name) for name in sorted((f for f in os.listdir(folder) if os.path.isdir(os.path.join(folder, f)) and folder_pattern.match(f)),
+                                                                       key=lambda x: int(x.split("part")[1]))):
+        node_feats_data.append(dgl.data.utils.load_tensors(f'{path}/node_feat.dgl'))
+
+    num_parts = len(node_feats_data)
+    metadata = {}
+    for f in feats:
+        print(f"Processing '{f}' features...")
+        if f not in node_feats_data[0]:
+            print(f"Error: Unknown feature '{f}'. Files contain the following features: {node_feats_data[0].keys()}.")
```
[the remark this answers is not in the document]
Comment (isratnisa) [reply tools/convert_feat_to_wholegraph.py]: Done

File: tools/README.md (hunk 5 above)
Comment (isratnisa) [reply tools/README.md]: Not exposing any names with the new design where it iterates through all the node feature names by default.

File: tools/README.md [hunk 9]
```diff
@@ -83,3 +83,51 @@ python3 /graphstorm/tools/partition_graph.py --dataset ogbn-arxiv \
...
+
+Use the `convert_feat_to_wholegraph.py` script with `--dataset-path` pointing to the distDGL folder of partitions and `--feat-names` to the features you want to convert (e.g. '`paper/feat,author/feat,institution/feat`')
```
[the remark this answers is not in the document]
Comment (isratnisa) [reply tools/README.md]: Yes

File: tools/README.md (hunk 4 above)
Comment (isratnisa) [reply tools/README.md:131]: That's true! But I wonder if it has something to do with WholeGraph dataloading. If it expects all the files to be in the same folder where different nodes reads parts of it? 

The file naming convention requires you to provide `num_parts` which outputs `_part_0_of_`
```
filename = wgth.utils.get_part_file_name(feat.replace('/', '~'), part_num, num_parts)
```
And the loading is as follows:
```
feat_path = os.path.join(os.path.dirname(part_config), 'wholegraph', ntype + '~feat')
node_feat_wm_embedding.get_embedding_tensor().from_file_prefix(feat_path,
                                                                       part_count=num_parts)
```

Comment (chang-l) [reply tools/README.md:131]: Another thing is, when num_part changes, you don't need to re-run this conversion if using single binary file for wholegraph features 😄

Comment (isratnisa) [reply tools/README.md:131]: Thats true!!

File: tools/convert_feat_to_wholegraph.py [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+    node_feats_data = []
+    folder_pattern = re.compile(r"^part[0-9]+$")
+    for path in (os.path.join(folder, name) for name in sorted( \
+        (f for f in os.listdir(folder) if os.path.isdir(os.path.join(folder, f)) \
+        and folder_pattern.match(f)), key=lambda x: int(x.split("part")[1]))):
+        node_feats_data.append(dgl.data.utils.load_tensors(f'{path}/node_feat.dgl'))
+
+    num_parts = len(node_feats_data)
+    metadata = {}
+    masks = ["train_mask", "test_mask", "val_mask", "labels"]
```
Comment (classicsong) [inline tools/convert_feat_to_wholegraph.py:47]: During training, a user can specify its label field through --label-field.
If we are going to store lables with DistDGL graph, we need to ask users to give use the feature name of label_field.

Comment (classicsong) [reply tools/convert_feat_to_wholegraph.py:47]: Can we store labels as normal feature instead?

File: tools/convert_feat_to_wholegraph.py [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+
+    for feat in list(node_feats_data[0].keys()):
+        if feat.split('/')[1] not in masks:
+            print(f"Processing '{feat}' features...")
+            if feat not in node_feats_data[0]:
```
Comment (classicsong) [inline tools/convert_feat_to_wholegraph.py:52]: You get ```feat``` from ```list(node_feats_data[0].keys())```, Why this if statement can be true?

File: tools/convert_feat_to_wholegraph.py [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+                raise RuntimeError(f"Error: Unknown feature '{feat}'. Files contain " +
+                                   "the following features: {node_feats_data[0].keys()}.")
+            whole_feat_tensor = torch.concat(tuple(t[feat] for t in node_feats_data), dim=0)
+            metadata[feat] = {'shape': list(whole_feat_tensor.shape),
+                           'dtype': str(whole_feat_tensor.dtype)}
```
Comment (classicsong) [inline tools/convert_feat_to_wholegraph.py:57]: indent

File: tools/convert_feat_to_wholegraph.py [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+        os.rename(os.path.join(folder, f'part{part}', 'node_feat.dgl'),
+                  os.path.join(folder, f'part{part}', 'node_feat.dgl.bak'))
+        os.rename(os.path.join(folder, f'part{part}', 'new_node_feat.dgl'),
+                  os.path.join(folder, f'part{part}', 'node_feat.dgl'))
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--dataset-path', action='store', type=str, required=True)
+    # parser.add_argument('--feat-names', action='store', type=str, required=True)
```
Comment (classicsong) [inline tools/convert_feat_to_wholegraph.py:102]: Please remove this comment.

File: tools/README.md [hunk 14]
```diff
@@ -83,3 +83,51 @@ python3 /graphstorm/tools/partition_graph.py --dataset ogbn-arxiv \
...
+## Convert node features from distDGL format to WholeGraph format
+
+Use the `convert_feat_to_wholegraph.py` script with `--dataset-path` pointing to the distDGL folder of partitions. It will convert all the node features to WholeGraph compatible format.
+```
+python3 convert_feat_to_wholegraph.py --dataset-path ogbn-mag240m-2p
+```
+
+The script will create a new folder '`wholegraph`' under '`ogbn-mag240m-2p`' containing the WholeGraph input files and will trim the distDGL file `node_feat.dgl` in each partition and save a backup `node_feat.dgl.bak`.
+
+The features in those files can be loaded in memory via the WholeGraph API by mentioning the folder path and feature prefix (`<node_type>~<feat_name>`).
```
[the remark this answers is not in the document]
Comment (isratnisa) [reply tools/README.md]: Fixed.

File: tools/convert_feat_to_wholegraph.py (hunk 11 above)
Comment (isratnisa) [reply tools/convert_feat_to_wholegraph.py]: Right! It doesn't make sense anymore.

File: tools/convert_feat_to_wholegraph.py [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+    # swap 'node_feat.dgl' files
+    for part in range(num_parts):
+        os.rename(os.path.join(folder, f'part{part}', 'node_feat.dgl'), os.path.join(folder, f'part{part}', 'node_feat.dgl.bak'))
+        os.rename(os.path.join(folder, f'part{part}', 'new_node_feat.dgl'), os.path.join(folder, f'part{part}', 'node_feat.dgl'))
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--dataset-path', action='store', type=str, required=True)
+    parser.add_argument('--feat-names', action='store', type=str, required=True)
```
[the remark this answers is not in the document]
Comment (isratnisa) [reply tools/convert_feat_to_wholegraph.py]: Where? Why is it necessary?

File: tools/README.md [hunk 16]
```diff
@@ -83,3 +83,51 @@ python3 /graphstorm/tools/partition_graph.py --dataset ogbn-arxiv \
...
+    "cpu",
+    "readonly", # access type
+    0.0, # cache ratio
+)
+
+paper_feat_wg = wgth.create_embedding(
+                    global_comm,
+                    'distributed',
+                    'cpu',
+                    getattr(torch, metadata['paper/feat']['dtype'].split('.')[1]),
```
[the remark this answers is not in the document]
Comment (chang-l) [reply tools/README.md:125]: Yes, it's here: [_pylibwholegraph.torch.utils.torch_dtype_to_wholememory_dtype()_](https://github.com/rapidsai/wholegraph/blob/d482448e967385b540c896484c506d246186ad6c/python/pylibwholegraph/pylibwholegraph/torch/utils.py#L22).
But in the current integration, we don't need wholegraph dtype (`wgth.create_embedding` only accepts torch dtype).

File: tools/convert_feat_to_wholegraph.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (isratnisa) [reply tools/convert_feat_to_wholegraph.py]: Update: The script will convert both features and labels. So, essentially it will convert all the attributes in `node_feat.dgl` except for `train_mask`, `test_mask`, and `val_mask`, as we can only guarantee the naming of these attributes. Labels and features can have any name and will be appropriately named during the conversion process.

File: tools/convert_feat_to_wholegraph.py (hunk 10 above)
Comment (isratnisa) [reply tools/convert_feat_to_wholegraph.py]: Done.

**Review by classicsong — APPROVED** [review]
LGTM
